tests/models/test_taskinstance.py

```diff
@@ -1347,34 +1361,38 @@ def test_check_task_dependencies(
             set_teardown, expect_state = expect_state
             assert isinstance(set_teardown, bool)
 
-        monkeypatch.setattr(_UpstreamTIStates, "calculate", lambda *_: upstream_states)
-
         # sanity checks
-        s = upstream_states
-        assert s.skipped >= s.skipped_setup
-        assert s.success >= s.success_setup
-        assert s.done == s.failed + s.success + s.removed + s.upstream_failed + s.skipped
+        assert len(upstream_states) == upstream_setups + 5
 
         with dag_maker() as dag:
             downstream = EmptyOperator(task_id="downstream", trigger_rule=trigger_rule)
             if set_teardown:
                 downstream.as_teardown()
+            work_tasks = []
             for i in range(5):
                 task = EmptyOperator(task_id=f"work_{i}", dag=dag)
                 task.set_downstream(downstream)
+                work_tasks.append(task)
             for i in range(upstream_setups):
                 task = EmptyOperator(task_id=f"setup_{i}", dag=dag).as_setup()
-                task.set_downstream(downstream)
+                task.set_downstream(work_tasks)
```

Comment:
Note: I re-modelled the test DAG such that the setup tasks are in front of the worker tasks, else the setup is directly linked to downstream teardown which (1) does not make sense and (2) is explicitly not handled.

Reply:
i don't understand.  it might be best to make a separate PR for your test refactor without simultaneously changing the code.  then followup with the real code change (and the test changes required just for that will be easier to see)